goss-testing/tests/ncn/goss-switch-ping-status.yaml

```diff
@@ -0,0 +1,42 @@
+#
+# MIT License
+#
+# (C) Copyright 2022 Hewlett Packard Enterprise Development LP
+#
+# Permission is hereby granted, free of charge, to any person obtaining a
+# copy of this software and associated documentation files (the "Software"),
+# to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense,
+# and/or sell copies of the Software, and to permit persons to whom the
+# Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included
+# in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+{{ $scripts := .Env.GOSS_BASE | printf "%s/scripts" }}
+{{ $logrun := $scripts | printf "%s/log_run.sh" }}
+command:
+    {{ $testlabel := "switch-ping-status" }}
+    {{$testlabel}}:
+        title: Verify that the network switch has external access
+        meta:
+            desc: If this test fails, then validate the network switch configurations.
+            sev: 0
+        exec: |-
+            #Set pipefail so that if any commands in the command pipeline fail, the test will fail
+            set -eo pipefail
+            "{{$logrun}}" --no-args -l "{{$testlabel}}" \
+                canu test --csm 1.2 --json --username admin --password ${SW_ADMIN_PASSWORD} |grep ping
+        stdout:
+          - "PASS"
```

Comment:
When I run this canu test manually on a system, this is the output I see:

```json
{
    "sw-spine-001": {
        "Software version test": "PASS",
        "lacp interfaces test": "PASS",
        "interface speed mismatch check": "FAIL",
        "VSX status check": "PASS",
        "VSX ISL check": "PASS",
        "VSX keepalive check": "PASS",
        "Interface error check": "PASS",
        "running-config different from startup-config": "PASS",
        "STP check for blocked ports": "PASS",
        "CPU Utilization over 70%": "PASS",
        "Memory Utilization over 70%": "PASS",
        "vlan 1 ip-helper test": "PASS",
        "vlan 2 ip-helper test": "PASS",
        "KEA connection check": "PASS",
        "vlan 4 ip-helper test": "PASS",
        "tftp route": "PASS",
        "BGP Test": "PASS",
        "STP check for root bridge spine": "PASS",
        "Is internal DNS server alive (ping)": "PASS",
        "Does internal DNS resolve ncn-m001-nmn": "PASS",
        "Internet connection check": "PASS"
    },
    "sw-spine-002": {
        "Software version test": "PASS",
        "lacp interfaces test": "PASS",
        "interface speed mismatch check": "FAIL",
        "VSX status check": "PASS",
        "VSX ISL check": "PASS",
        "VSX keepalive check": "PASS",
        "Interface error check": "PASS",
        "running-config different from startup-config": "PASS",
        "STP check for blocked ports": "PASS",
        "CPU Utilization over 70%": "PASS",
        "Memory Utilization over 70%": "PASS",
        "vlan 1 ip-helper test": "PASS",
        "vlan 2 ip-helper test": "PASS",
        "KEA connection check": "PASS",
        "vlan 4 ip-helper test": "PASS",
        "tftp route": "PASS",
        "BGP Test": "PASS",
        "STP check for root bridge spine": "PASS",
        "Is internal DNS server alive (ping)": "PASS",
        "Does internal DNS resolve ncn-m001-nmn": "PASS",
        "Internet connection check": "PASS"
    },
    "sw-leaf-bmc-001": {
        "Software version test": "PASS",
        "lacp interfaces test": "PASS",
        "interface speed mismatch check": "PASS",
        "Interface error check": "PASS",
        "running-config different from startup-config": "PASS",
        "STP check for blocked ports": "PASS",
        "CPU Utilization over 70%": "PASS",
        "Memory Utilization over 70%": "PASS",
        "KEA connection check": "PASS",
        "STP check for root bridge leaf": "PASS",
        "Is internal DNS server alive (ping)": "PASS",
        "Does internal DNS resolve ncn-m001-nmn": "PASS",
        "Internet connection check": "PASS"
    }
}
```

If you grep the output for ping, you get:
```text
        "Is internal DNS server alive (ping)": "PASS",
        "Is internal DNS server alive (ping)": "PASS",
        "Is internal DNS server alive (ping)": "PASS",
```

The problem with just checking for PASS is that this means the Goss test will pass if even a single switch reports that it passed, even if the others report failure. The test needs to not only verify that it sees passes (to make sure that the tests actually ran and succeeded), but also verify that no failures show up. 

Really, since the output is in JSON format anyway, it would probably be better to write a small Python script to parse it and validate these things. It may be possible to do it with jq, but I'm not sure.